### DISCUSSION TITLE
Delete origin/<branch> on the remote (not just locally)?

### DISCUSSION THREAD

**Discussion by phil-s** [magit/magit#4905] [Ideas]
I went to delete a remote branch via the `magit-refs-mode` buffer, and was prompted thus:

    Delete origin/<branch> on the remote (not just locally)? (y or n)

That "just" is confusing and caused me concern, as it made it look like Magit was intending to remove the local branch, and was then asking me whether I wanted to *additionally* remove the remote branch.  It did the right thing, but I think that prompt can be clarified.  I suggest this:

    Delete origin/<branch> on the remote (not locally)? (y or n)

Magit v3.3.0-dirty, Git 2.17.1, Emacs 28.2, gnu/linux

**Comment by tarsius** [comment]
You are not the first to find this confusing (though I cannot find the other report), so I have added additional information in 781544586b0b69ddff193b65d5d0f9c923a385b2, to point you towards

>  `delete-branch-on-remote` Deleting a \"remote branch\" may mean
>  deleting the (local) \"remote-tracking\" branch only, or also
>  removing it from the remote itself.  The latter often makes more
>  sense because otherwise simply fetching from the remote would
>  restore the remote-tracking branch, but doing that can be
>  surprising and hard to recover from, so we ask.

When you answer *no* at that prompt, then only the local *remote-tracking* branch is deleted, e.g., the reference `refs/remotes/origin/feature`. That is a ref in your local repository, which tracks the state of a branch on the remote. Deleting it is a local operation. This is done using `git update-ref -d refs/remotes/origin/feature`.

If you answer *yes*, then the branch is *also* deleted *on the remote* using something like `git push origin :feature`.

**Reply by phil-s** [reply]
Ah, that makes much more sense now!

How about this prompt:

    Deleting local refs/remotes/origin/<branch>.  Also delete origin/<branch> on the remote? (y or n)

**Reply by tarsius** [reply]
```
Deleting local refs/remotes/origin/<branch>.  Also delete <branch> on origin? (y or n)
```

Seems more appropriate if we go with a more technical description because there isn't actually a `origin/<branch>` on the remote, which could be deleted. It is likely called `refs/heads/<branch>` there.

I realize the current prompt contains the same inaccuracy, but there it isn't in contrast with using the full ref to name the thing being deleted locally. I think it is unfortunate that `git`, when dealing with a branch located on a remote (fetching/pushing/...), requires the remote and branch to be given as two separate arguments, as in `git fetch origin branch`, but for everything else `<remote>/<branch>` is acceptable. Magit for the most part ignores this inconsistency in `git`'s interface and automatically translates to the wire-form when required.

I am not sure we should change the prompt like this. I feel putting the full ref in there, might just confuse a different set of users. @kyleam what do you think?

**Reply by phil-s** [reply]
> might just confuse a different set of users

That's certainly possible.

Perhaps this prompt actually needs a additional `(h)elp` option, so that users can easily get a detailed explanation.

**Reply by kyleam** [reply]
> I am not sure we should change the prompt like this. I feel putting the full ref in there, might just confuse a different set of users. @kyleam what do you think?

I'm in favor of your last suggestion ("Deleting local refs/remotes/origin/<branch>.  Also delete <branch> on origin? (y or n)").

I think it's clearer and will cause less confusion overall.  It seems to me there are two sources of confusion: 1) confusion about Magit's behavior (what prompted this discussion) and 2) general confusion over the concept of remote-tracking branches.  The proposed wording change helps with the first, and I think spelling out the full ref will help people that don't know about remote-tracking branches track down the info they need.

**Reply by tarsius** [reply]
I've gone with the full-ref approach for now. The (h)elp idea isn't bad either; maybe we'll do that eventually, here and in other places.

**Comment by phil-s** [comment]
Looking at the commit https://github.com/magit/magit/commit/b7daf6a2ff00f7752d05dec17f2f5a825954b151 the PROMPT-N argument immediately below still has the "not just locally" phrasing:

    "Delete %i branches on the remote (not just locally)"

How about this?

```diff
          ((magit-confirm 'delete-branch-on-remote
-            "Delete %s on the remote (not just locally)"
-            "Delete %i branches on the remote (not just locally)"
-            'noabort branches)
+            (format "Deleting local %s.  Also delete on %s"
+                    (magit-ref-fullname (car branches))
+                    remote)
+            (format "Deleting %i local refs.  Also delete on %s"
+                    (length refs)
+                    remote)
+            'noabort refs)
           ;; The ref may actually point at another rev on the remote,
```

**Reply by tarsius** [reply]
That is problematic when a user doesn't know that "remote-tracking branches" are "local refs". Of course we could say, "local remote-tracking branches", but then many would wonder how something can be both local and remote at the same time.

**Reply by phil-s** [reply]
I thought it analogous to the initial change.  The new UI for that is firstly:
```
Delete branch (default origin/phil/foo1): 
```
followed by:
```
Deleting local refs/remotes/origin/phil/foo1.  Also delete on origin? (y or n) 
```
And my suggested change for multiple deletions gives us firstly:
```
origin/phil/foo1
origin/phil/foo2
origin/phil/foo3

Delete 3 branches? (y or n) 
```
followed by:
```
refs/remotes/origin/phil/foo1
refs/remotes/origin/phil/foo2
refs/remotes/origin/phil/foo3

Deleting 3 local refs.  Also delete on origin? (y or n) 
```

**Reply by tarsius** [reply]
Ah right. Yes, that would make sense. For people like me it would be needlessly verbose, but there's not much we can do about that.

**Reply by phil-s** [reply]
Shall we go ahead with that?

**Reply by tarsius** [reply]
Yes, sounds good.

**Reply by phil-s** [reply]
I've opened https://github.com/magit/magit/pull/4914 for that.
